Auto create data streams using index templates v2

martijnvg opened the pull request:
This PR adds the ability to auto create data streams using index templates v2.
Index templates (v2) now have a data_steam field that includes a timestamp field,
if provided and index name matches with that template then a data stream
(plus first backing index) is auto created. The index/bulk apis will redirect
to the create data stream api instead of the create index api.

Relates to #53100

martijnvg committed:
WIP commit

This commit adds the ability to auto create data streams using index templates v2.
Index templates (v2) now have a data_steam field that includes a timestamp field,
if provided and index name matches with that template then a data stream
(plus first backing index) is auto created. The index/bulk apis will redirect
to the create data stream api instead of the create index api.

Relates to #53100

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

martijnvg committed:
fixed hlrc test compile errors

martijnvg committed:
do not serialize data stream template to nodes that don't yet know how to de-serialize it

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

martijnvg committed:
fixed issues after merge

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

martijnvg committed:
added tests

martijnvg committed:
rename

martijnvg committed:
fixed test compile errors and
removed unused imports

danhermann commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.data_stream/30_auto_create_data_stream.yml:
```diff
@@ -0,0 +1,51 @@
+      allowed_warnings:
+        - "index template [test] has index patterns [test-*] matching patterns from existing older templates [global] with patterns (global => [*]); this template [test] will take precedence during new index creation"
+      indices.put_index_template:
+        name: generic_logs_template
```
Do we not need to remove this ITv2 at the end of the test?

dakrone replied:
V2 templates are removed at the end of each test automatically, so shouldn't be a concern

danhermann commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -272,6 +296,64 @@ public void onFailure(Exception e) {
+                }
+            };
+            Consumer<Exception> onFailure = e -> {
+                if (!(ExceptionsHelper.unwrapCause(e) instanceof ResourceAlreadyExistsException)) {
+                    // fail all requests involving this index, if create didn't work
+                    for (int i = 0; i < bulkRequest.requests.size(); i++) {
+                        DocWriteRequest<?> item = bulkRequest.requests.get(i);
+                        if (item != null && setResponseFailureIfIndexMatches(responses, i, item, name, e)) {
+                            bulkRequest.requests.set(i, null);
+                        }
+                    }
+                }
```
Duplicates code in the `onFailure` handler in the `doExecute` method. Might be worth consolidating.

danhermann commented on server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionTests.java:
```diff
@@ -287,4 +298,44 @@ public void testResolveRequestOrDefaultPipelineAndFinalPipeline() {
+        Map<String, DataStreamTemplate> result = TransportBulkAction.resolveAutoCreateDataStreams(metadata, Set.of());
+        assertThat(result, anEmptyMap());
+
+        Set<String> autoCreateIndices = new HashSet<>(Set.of("logs-foobar", "logs-barbaz"));
```
Would this variable be better named `autoCreateDataStreams`?

danhermann reviewed (approved):
LGTM. Left some questions about minor, non-essential items.

henningandersen commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -272,6 +296,64 @@ public void onFailure(Exception e) {
+        Iterator<String> autoCreateIndicesIterator = autoCreateIndices.iterator();
+        while (autoCreateIndicesIterator.hasNext()) {
+            String indexName = autoCreateIndicesIterator.next();
+            String v2Template = MetadataIndexTemplateService.findV2Template(metadata, indexName, false);
```
I think this means that if you have both v1 and v2 templates matching the same pattern and the v2 one is a data stream, it takes precedence even when `prefer_v2_templates=false`. I think this will cause some confusion and I would prefer to stick to respecting the option.

henningandersen reviewed (changes requested):
Thanks @martijnvg. I added a few comments that I believe need to be addressed.

martijnvg commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -272,6 +296,64 @@ public void onFailure(Exception e) {
         }
     }
 
+    static Map<String, IndexTemplateV2.DataStreamTemplate> resolveAutoCreateDataStreams(Metadata metadata,
```
That is trappy indeed, thanks for spotting this. I will update the PR, so that most of the auto create logic here is moved to the to be built auto create action.

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

martijnvg committed:
fixed compile errors after merge

martijnvg committed:
extracted the create data stream logic from CreateDataStreamAction.TransportAction to
MetadataCreateDataStreamService class. This new class can then also be used from
the to be created auto create transport action.

martijnvg committed:
introduced auto create action.

For index and bulk requests that refer to indices/data streams that don't exist yet,
the TransportBulkAction will redirect to the auto create action which will either
create an index or data stream.

Previously this decision was made in the TransportBulkAction,
on the coordinating node, but if that node is currently unaware of the existence of
specific data streams it will create an index instead of a data stream.

martijnvg commented:
After adding the auto create action, the size of the PR increased too much to be able to do a careful review. Also the PRs making two different changes. So I've split the auto create action in a new pr (#55858). I will update this pr after that PR gets merged.

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

martijnvg committed:
fixed post merge issues

martijnvg committed:
Decide whether a data stream or index should be created on
the elected master node in the auto create transport action.

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

martijnvg committed:
fixed compile errors after merge

martijnvg commented:
I've updated this PR now that #56122 has been merged.

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

martijnvg committed:
fixed test compile error after merging in master

henningandersen commented on test/framework/src/main/java/org/elasticsearch/test/ClusterServiceUtils.java:
```diff
@@ -100,8 +100,12 @@ public void onFailure(String source, Exception e) {
+        return createClusterService(threadPool, Version.CURRENT);
+    }
+
+    public static ClusterService createClusterService(ThreadPool threadPool, Version nodeVersion) {
```
I am not sure this is used?

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateV2.java:
```diff
@@ -88,13 +95,15 @@ public static IndexTemplateV2 parse(XContentParser parser) throws IOException {
     public IndexTemplateV2(List<String> indexPatterns, @Nullable Template template, @Nullable List<String> componentTemplates,
-                           @Nullable Long priority, @Nullable Long version, @Nullable Map<String, Object> metadata) {
+                           @Nullable Long priority, @Nullable Long version, @Nullable Map<String, Object> metadata,
+                           @Nullable DataStreamTemplate dataStreamTemplate) {
```
Since this is part of the HLRC, I wonder if we should stay backwards compatible with the original constructor, i.e., have both variants? Not sure of the exact policy on that subject though.

martijnvg replied:
I think changing this constructor now is okay, because index templates v2 is experimental in 7.8? I don't think we have a real policy regarding hlrc bwc.

martijnvg replied:
I added the constructor:  https://github.com/elastic/elasticsearch/pull/55377/commits/5ed50fe948a24505257bfcae8d4226037a4e399a

henningandersen reviewed (changes requested):
Thanks for the extra iteration. I have a few more comments, but otherwise this is looking good.

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

martijnvg commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:
```diff
@@ -84,4 +107,21 @@ protected ClusterBlockException checkBlock(CreateIndexRequest request, ClusterSt
+
+    private static boolean resolvePreferV2Templates(CreateIndexRequest request) {
+        return request.preferV2Templates() == null ?
+            IndexMetadata.PREFER_V2_TEMPLATES_SETTING.get(request.settings()) : request.preferV2Templates();
```
True, the settings in the request are always empty.  However via this way we can ensure that we always rely on prefer_v2_templates default, also when this change gets backported to 7.x.
Should we use `Settings.EMPTY` here instead of `request.settings()`?

henningandersen replied:
Sure, that makes the expectation a bit clearer. Probably a good idea to add a comment too.

martijnvg committed:
applied smaller feedback items

martijnvg committed:
overload constructor

martijnvg commented on server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:
```diff
@@ -75,7 +85,20 @@ protected void masterOperation(Task task,
                                        ClusterState state,
                                        ActionListener<CreateIndexResponse> listener) {
-            TransportCreateIndexAction.innerCreateIndex(request, listener, indexNameExpressionResolver, createIndexService);
+            DataStreamTemplate dataStreamTemplate = resolveAutoCreateDataStream(request, state.metadata());
```
Let's do this check and auto creating an index or data stream in a single cluster state update. Then we don't need to verify after the fact that an index or data stream has been created.

martijnvg committed:
Perform the check whether a data stream or index should be auto created in the same update
that creates the index or data stream.

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:
```diff
@@ -0,0 +1,119 @@
+
+    public void createDataStream(CreateDataSteamClusterStateUpdateRequest request,
+                                 ActionListener<AcknowledgedResponse> listener) {
+        clusterService.submitStateUpdateTask("create-data-stream [" + request.name + "]",
```
Yes, tests could fail if a coordinating/data node has a stale cluster state for too long.

martijnvg committed:
wait for active shards when auto creating data streams / indices and
when creating data streams via api

martijnvg committed:
undo style changes

martijnvg committed:
undo other style changes

martijnvg committed:
undo other style changes

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:
```diff
@@ -0,0 +1,148 @@
+            response -> {
+                if (response.isAcknowledged()) {
+                    String firstBackingIndexName = firstBackingIndexRef.get();
+                    assert finalListener != null;
```
typo?
```suggestion
                    assert firstBackingIndexName != null;
```

henningandersen reviewed (approved):
LGTM.

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

martijnvg commented on server/src/internalClusterTest/java/org/elasticsearch/action/bulk/BulkIntegrationIT.java:
```diff
@@ -200,4 +217,104 @@ public void testDeleteIndexWhileIndexing() throws Exception {
+
+        Boolean preferV2Templates = randomBoolean() ? null : true;
+        if (randomBoolean()) {
+            PutIndexTemplateRequest v1Request = new PutIndexTemplateRequest("logs-foo");
```
I will remove this randomization now that the prefer flag has been removed.

martijnvg committed:
iter

martijnvg committed:
fixed test failure

martijnvg committed:
adjust serialization logic

martijnvg committed:
Merge remote-tracking branch 'es/master' into auto_create_ds_with_itv2

martijnvg committed:
fixed checkstyle

martijnvg committed:
fixed serialization order
